devito/ir/clusters/algorithms.py

```diff
@@ -1,75 +1,182 @@
-from collections import OrderedDict, namedtuple
+from collections import OrderedDict
 
-from devito.ir.clusters.cluster import Cluster
+from devito.ir.support import Scope
+from devito.ir.clusters.cluster import PartialCluster, ClusterGroup
 from devito.ir.dfg import TemporariesGraph
-from devito.ir.support import Stencil
 from devito.symbolics import xreplace_indices
 from devito.types import Scalar
+from devito.tools import flatten
 
-__all__ = ['clusterize', 'optimize']
+__all__ = ['clusterize', 'groupby']
 
 
-def merge(clusters):
+def groupby(clusters):
     """
-    Given an ordered collection of :class:`Cluster` objects, return a
-    (potentially) smaller sequence in which clusters with identical stencil
-    have been merged into a single :class:`Cluster`.
+    Given an ordered collection of :class:`PartialCluster` objects, return a
+    (potentially) smaller sequence in which dependence-free PartialClusters with
+    identical stencil have been squashed into a single PartialCluster.
     """
-    mapper = OrderedDict()
+    clusters = clusters.unfreeze()
+
+    # Attempt cluster fusion
+    processed = ClusterGroup()
     for c in clusters:
-        mapper.setdefault((c.stencil.entries, c.atomics), []).append(c)
+        fused = False
+        for candidate in reversed(list(processed)):
+            # Check all data dependences relevant for cluster fusion
+            scope = Scope(exprs=candidate.exprs + c.exprs)
+            anti = scope.d_anti.carried() - scope.d_anti.increment
+            flow = scope.d_flow - (scope.d_flow.inplace() + scope.d_flow.increment)
+            funcs = [i.function for i in anti]
+            if candidate.stencil == c.stencil and\
+                    all(is_local(i, candidate, c, clusters) for i in funcs):
+                # /c/ will be fused into /candidate/. All fusion-induced anti
+                # dependences are eliminated through so called "index bumping and
+                # array contraction", which transforms array accesses into scalars
```

Comment:
Out of interest, is there a book or reference for all the technical compiler lingo that our backends now use?

Reply:
this is pretty much standard compiler jargon, you can find it in lecture notes (see for example: http://www.cs.cmu.edu/afs/cs/academic/class/15745-s06/web/handouts/15.pdf, http://www.cs.utah.edu/~mhall/cc/Utah-L1.pdf, ...) or in any textbook, such as the "Dragon Book " (https://en.wikipedia.org/wiki/Compilers:_Principles,_Techniques,_and_Tools)

I think/hope I haven't invented anything new except for things that are special to us